imap/index.c

```diff
@@ -3813,6 +3813,124 @@ static int index_fetchannotations(struct index_state *state,
     return r;
 }
 
+struct fetch_mailbox_rock {
+    struct index_state *state;
+    int sep;
+    int wantname;
+    const struct fetchargs *fetchargs;
+};
+
+static int fetch_mailbox_cb(const conv_guidrec_t *rec, void *rock)
+{
+    static const int needrights = ACL_READ|ACL_LOOKUP;
+    struct fetch_mailbox_rock *fmb_rock = (struct fetch_mailbox_rock *) rock;
+    mbentry_t *mbentry = NULL;
+    int myrights = 0;
+    struct mailbox *mailbox = NULL;
+    msgrecord_t *msgrecord = NULL;
+    char *extname = NULL;
+    uint32_t system_flags, internal_flags;
+    int r = 0;
+
+    assert(fmb_rock->state != NULL);
+    assert(fmb_rock->fetchargs != NULL);
+
+    /* make sure we have appropriate rights */
+    r = mboxlist_lookup(rec->mboxname, &mbentry, NULL);
+    if (r) goto done;
+    myrights = cyrus_acl_myrights(fmb_rock->state->authstate, mbentry->acl);
+    if ((myrights & needrights) != needrights)
+        goto done;
+
+    if (rec->version >= 1) {
+        /* grab flags from convdb, if it has them */
+        system_flags = rec->system_flags;
+        internal_flags = rec->internal_flags;
+    }
+    else {
+        /* grab flags from message record */
```

Comment:
Opening the mailbox in the GUID callback will be expensive if clients fetch a sequence set of non-trivial length. I don't think this will be an issue, given that opening the mailbox here is just a fallback for old conversation.dbs. But generally we suffered in the JMAP implementation at quite some places and turned to handling all messages in the same mailbox in bulk.

Reply:
I think I agree that this won't be an issue --  it will only affect people who are running the bleeding edge code (and get this feature) but haven't updated their convdb yet.

Once this becomes general release, presumably people will update their convdb as part of their upgrade process, and will just get the cheap version immediately.